File: java/src/main/java/money_problem/domain/Portfolio.java
package money_problem.domain;

import io.vavr.collection.List;

public record Portfolio(Money... moneys) {
    private static final double EURO_TO_USD = 1.2;

    public Money evaluate(Currency currency) {
        return new Money(List.of(moneys)
                .foldLeft(0d, (acc, money) -> acc + convert(money, currency)), currency);
    }

    private static double convert(Money money, Currency currency) {
        return currency == money.currency()
                ? money.amount()
                : money.amount() * EURO_TO_USD;
    }
}

File: kotlin/src/main/kotlin/money_problem/domain/Portfolio.kt
package money_problem.domain

import arrow.core.Either

data class Portfolio(val moneys: List<Money>)

fun Portfolio.evaluate(
    bank: Bank,
    toCurrency: Currency
): Either<String, Money> {
    val convertedMoneys = moneys.map { bank.convert(it, toCurrency) }
    val lefts = convertedMoneys.lefts()

    return if (lefts.isEmpty())
        Either.Right(
            Money(
                convertedMoneys
                    .rights()
                    .fold(0.0) { acc, money -> acc + money.amount }, toCurrency
            )
        )
    else Either.Left(
        lefts.joinToString(separator = ",", prefix = "Missing exchange rate(s): [", postfix = "]")
    )
}